src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/threadpool/QueueRejectionRca.java

```diff
@@ -0,0 +1,190 @@
+            nodeSummary = new HotNodeSummary(currentNode.getId(), currentNode.getHostAddress());
+          }
+          nodeSummary.appendNestedSummary(collector.generateSummary(currTimestamp));
+        }
+      }
+      ResourceContext context;
+      if (nodeSummary == null) {
+        context = new ResourceContext(Resources.State.HEALTHY);
+      }
+      else {
+        context = new ResourceContext(Resources.State.UNHEALTHY);
+      }
+      boolean isDataNode = !currentNode.getIsMasterNode();
+      return new ResourceFlowUnit<>(currTimestamp, context, nodeSummary, isDataNode);
+    }
+    else {
+      return new ResourceFlowUnit<>(currTimestamp);
+    }
+  }
+
+  // TODO: move this method back into the Rca base class
+  @Override
+  public void generateFlowUnitListFromWire(FlowUnitOperationArgWrapper args) {
+    final List<FlowUnitMessage> flowUnitMessages =
+        args.getWireHopper().readFromWire(args.getNode());
+    List<ResourceFlowUnit<HotNodeSummary>> flowUnitList = new ArrayList<>();
+    LOG.debug("rca: Executing fromWire: {}", this.getClass().getSimpleName());
+    for (FlowUnitMessage flowUnitMessage : flowUnitMessages) {
+      flowUnitList.add(ResourceFlowUnit.buildFlowUnitFromWrapper(flowUnitMessage));
+    }
+    setFlowUnits(flowUnitList);
+  }
+
+  /**
+   * A collector class to collect rejection from each queue type
+   */
+  private static class QueueRejectionCollector {
+    private final ResourceType threadPool;
+    private final ThreadPoolType threadPoolMetric;
+    private final Metric threadPool_RejectedReqs;
+    private boolean hasRejection;
+    private long rejectionTimestamp;
+    private long rejectionTimePeriodThreshold;
+
+    public QueueRejectionCollector(final ThreadPoolEnum threadPool, final ThreadPoolType threadPoolMetric,
+        final Metric threadPool_RejectedReqs, final long threshold) {
+      this.threadPool = ResourceType.newBuilder().setThreadPool(threadPool).build();
+      this.threadPoolMetric = threadPoolMetric;
+      this.threadPool_RejectedReqs = threadPool_RejectedReqs;
+      this.hasRejection = false;
+      this.rejectionTimestamp = 0;
+      this.rejectionTimePeriodThreshold = threshold;
+    }
+
+    public void collect(final long currTimestamp) {
+      for (MetricFlowUnit flowUnit : threadPool_RejectedReqs.getFlowUnits()) {
+        if (flowUnit.isEmpty()) {
+          continue;
+        }
+        double rejectCnt = SQLParsingUtil.readDataFromSqlResult(flowUnit.getData(),
+            THREAD_POOL_TYPE.getField(), threadPoolMetric.toString(), MetricsDB.MAX);
+        if (!Double.isNaN(rejectCnt)) {
+          if (rejectCnt > 0) {
+            if (!hasRejection) {
+              rejectionTimestamp = currTimestamp;
+            }
+            hasRejection = true;
+          }
+          else {
+            hasRejection = false;
+          }
+        }
+        else {
+          LOG.error("Failed to parse metric from threadpool {}", threadPool.toString());
```

Comment:
If we fail to parse metrics from the threadpool, should we be safe and reset hasrejection ? Or else, in case, if the rejection was reported and then all subsequent values are Nan for the rest of the 5 minutes, we will still change the queue size.